### PR TITLE
rdar://38123690 (overlarge batches) and rdar://37865437 (task queue exec errors)

### PR DESCRIPTION
**Pull request by graydon**
This fixes another assert @davidungar ran into testing batch mode, this time due to overlong command lines / overlarge batches (to be fixed when we move to passing OFMs, but not fixed yet). For the time being we just cap at total/TOO_MANY_FILES, which _might_ still not be small enough given the auxiliary file amplification, but it's enough to pass a credible test, and none of this is especially precise in the existing code either (I think the actual limit arises from colliding with ARG_MAX, which we're definitely not measuring!)

Along the way, fixes (by necessity) the cause of an assert @ddunbar hit, wherein the driver doesn't trap exec-failure properly; there's not a _lot_ we can recover in these cases, but we can do better than throwing an unrelated assert (or not trapping at all in release builds). The overlong args were triggering exec failure, but then so do all sorts of things, such as exec'ing a nonexistent binary.

rdar://37865437
rdar://38123690

### CODE REVIEW

**Comment by graydon** [comment]
@swift-ci please test

File: include/swift/Driver/ToolChain.h [hunk 1]
```diff
@@ -29,6 +29,9 @@ namespace driver {
   class Job;
   class OutputInfo;
 
+/// The limit for passing a list of files on the command line.
+const size_t TOO_MANY_FILES = 128;
```
Comment (jrose-apple) [inline include/swift/Driver/ToolChain.h]: Please make this `static`, there's no need for it to have a stable address.

File: lib/Driver/Compilation.cpp [hunk 2]
```diff
@@ -810,15 +814,29 @@ namespace driver {
...
+                                  _1, _2, _3, _4, _5),
+                        std::bind(&PerformJobsState::taskSignalled, this,
+                                  _1, _2, _3, _4, _5, _6))) {
+          if (Result == EXIT_SUCCESS) {
+            // Error from task queue while Result == EXIT_SUCCESS most likely
+            // means some fork/exec or posix_spawn failed; TaskQueue saw "an
+            // error" at some stage before even calling us with a process exit /
+            // signal (or else a poll failed); unfortunately the task causing it
+            // was dropped on the floor and we have no way to recover it here,
+            // so we report a very poor, generic error.
```
Comment (jrose-apple) [inline lib/Driver/Compilation.cpp]: Better than nothing, and I'm glad we caught it. Maybe this should be a FIXME, though?

**Review by jrose-apple — COMMENTED** [review]
Hmm…I'm not convinced TOO_MANY_FILES is good enough. Previously we had N inputs and N primary outputs but now we also have N of each supplementary output. I guess it's better if we have something that works in more circumstances right now, though.

File: lib/Driver/Compilation.cpp (hunk 2 above)
Comment (graydon) [reply lib/Driver/Compilation.cpp]: Sure

File: include/swift/Driver/ToolChain.h (hunk 1 above)
Comment (graydon) [reply include/swift/Driver/ToolChain.h]: Ok

**Comment by graydon** [comment]
@jrose-apple I'm not convinced TOO_MANY_FILES is great either; have rewritten this with a slightly-more-meaningful test of whether the overall command-line length is too long, with a rebatching cycle (at twice the number of partitions / half the batch size) if it fails.

**Comment by graydon** [comment]
@swift-ci please test

**Comment by graydon** [comment]
@swift-ci please test

**Review by jrose-apple — APPROVED** [review]

**Pull request merged**